File: src/v2/validation/schema.ts
/**
 * Schema Validation Module
 *
 * Config-driven structure validation for rule logic.
 * Uses ValidationConfig from workflow config to determine valid types.
 */

import { ValidationConfig } from '../config/types.js';

// =============================================================================
// TYPES
// =============================================================================

export interface ValidationResult {
  valid: boolean;
  errors: string[];
  warnings?: string[];
}

// =============================================================================
// SCHEMA VALIDATOR
// =============================================================================

/**
 * Validate rule logic structure against config-driven rules
 */
export function validateRuleLogic(
  logic: unknown,
  validationConfig: ValidationConfig
): ValidationResult {
  const errors: string[] = [];
  const warnings: string[] = [];

  if (!logic || typeof logic !== 'object') {
    return { valid: false, errors: ['Root must be an object'] };
  }

  const l = logic as Record<string, unknown>;

  // Required fields
  if (l.grammarVersion !== validationConfig.grammarVersion) {
    errors.push(
      `grammarVersion must be '${validationConfig.grammarVersion}', got: ${l.grammarVersion}`
    );
  }

  if (!l.workflowRid || typeof l.workflowRid !== 'string') {
    errors.push('workflowRid is required and must be a string');
  }

  // Strategy validation
  if (!l.strategy || typeof l.strategy !== 'object') {
    errors.push('strategy is required and must be an object');
  } else {
    const strategy = l.strategy as Record<string, unknown>;
    const strategyType = strategy.type as string;

    if (!validationConfig.supportedStrategyTypes.includes(strategyType)) {
      errors.push(
        `strategy.type must be one of [${validationConfig.supportedStrategyTypes.join(', ')}], got: ${strategyType}`
      );
    }

    // Check that the corresponding node exists
    if (strategyType && !strategy[strategyType]) {
      errors.push(`strategy.${strategyType} is required when type is '${strategyType}'`);
    }

    // Check for common mistakes in filterNode
    if (strategyType === 'filterNode') {
      const filterNode = strategy.filterNode as Record<string, unknown> | undefined;
      if (filterNode && 'type' in filterNode) {
        errors.push('filterNode should NOT have a type field (type goes in strategy)');
      }
    }
  }

  // Effect validation
  if (!l.effect || typeof l.effect !== 'object') {
    errors.push('effect is required and must be an object');
  } else {
    const effect = l.effect as Record<string, unknown>;
    if (effect.type !== 'v2') {
      errors.push(`effect.type must be 'v2', got: ${effect.type}`);
    }
    if (!effect.v2 || typeof effect.v2 !== 'object') {
      errors.push('effect.v2 is required');
    } else {
      const v2 = effect.v2 as Record<string, unknown>;
      if (!v2.outputAndVersion) {
        errors.push('effect.v2.outputAndVersion is required');
      }
    }
  }

  return {
    valid: errors.length === 0,
    errors,
    warnings: warnings.length > 0 ? warnings : undefined
  };
}

/**
 * Extract the strategy type from rule logic
 */
export function getStrategyType(logic: unknown): string | null {
  if (!logic || typeof logic !== 'object') return null;
  const l = logic as Record<string, unknown>;
  const strategy = l.strategy as Record<string, unknown> | undefined;
  return (strategy?.type as string) || null;
}

/**
 * Extract the object type ID from rule logic (works with any node type)
 */
export function extractObjectTypeId(logic: unknown): string | null {
  if (!logic || typeof logic !== 'object') return null;

  const l = logic as Record<string, unknown>;
  const strategy = l.strategy as Record<string, unknown> | undefined;

  if (!strategy) return null;

  // Try each node type
  const nodeTypes = ['filterNode', 'windowNode', 'aggregationNode'];

  for (const nodeType of nodeTypes) {
    const node = strategy[nodeType] as Record<string, unknown> | undefined;
    const source = (node?.nodeInput as Record<string, unknown>)?.source as Record<string, unknown>;
    if (source?.objectTypeId) {
      return source.objectTypeId as string;
    }
  }

  return null;
}

/**
 * Extract the workflow RID from rule logic
 */
export function extractWorkflowRid(logic: unknown): string | null {
  if (!logic || typeof logic !== 'object') return null;
  return (logic as Record<string, unknown>).workflowRid as string || null;
}

/**
 * Validate that workflowRid matches expected value
 */
export function validateWorkflowRid(
  logic: unknown,
  expectedRid: string
): ValidationResult {
  const errors: string[] = [];
  const actualRid = extractWorkflowRid(logic);

  if (!actualRid) {
    errors.push('workflowRid not found in rule logic');
  } else if (actualRid !== expectedRid) {
    errors.push(
      `workflowRid mismatch: expected '${expectedRid}', got '${actualRid}'`
    );
  }

  return { valid: errors.length === 0, errors };
}

/**
 * Validate that objectTypeId matches expected value
 */
export function validateObjectType(
  logic: unknown,
  expectedId: string
): ValidationResult {
  const errors: string[] = [];
  const actualId = extractObjectTypeId(logic);

  if (!actualId) {
    errors.push('objectTypeId not found in rule logic');
  } else if (actualId !== expectedId) {
    errors.push(
      `objectTypeId mismatch: expected '${expectedId}', got '${actualId}'`
    );
  }

  return { valid: errors.length === 0, errors };
}

File: src/foundry_rules/validation/schema.py
"""
Schema Validation Module

Config-driven structure validation for rule logic.
"""

from typing import Any, Optional
from pydantic import BaseModel

from ..config.types import ValidationConfig


class ValidationResult(BaseModel):
    """Result of validation."""

    valid: bool
    errors: list[str] = []
    warnings: list[str] = []


def validate_rule_logic(logic: Any, validation_config: ValidationConfig) -> ValidationResult:
    """
    Validate rule logic structure against config-driven rules.

    Args:
        logic: Rule logic dict
        validation_config: Validation configuration

    Returns:
        ValidationResult with any errors
    """
    errors: list[str] = []
    warnings: list[str] = []

    if not logic or not isinstance(logic, dict):
        return ValidationResult(valid=False, errors=["Root must be an object"])

    # Check grammar version
    if logic.get("grammarVersion") != validation_config.grammar_version:
        errors.append(
            f"grammarVersion must be '{validation_config.grammar_version}', "
            f"got: {logic.get('grammarVersion')}"
        )

    # Check workflowRid
    workflow_rid = logic.get("workflowRid")
    if not workflow_rid or not isinstance(workflow_rid, str):
        errors.append("workflowRid is required and must be a string")

    # Check strategy
    strategy = logic.get("strategy")
    if not strategy or not isinstance(strategy, dict):
        errors.append("strategy is required and must be an object")
    else:
        strategy_type = strategy.get("type")

        if strategy_type not in validation_config.supported_strategy_types:
            errors.append(
                f"strategy.type must be one of "
                f"[{', '.join(validation_config.supported_strategy_types)}], "
                f"got: {strategy_type}"
            )

        # Check that the corresponding node exists
        if strategy_type and strategy_type not in strategy:
            errors.append(
                f"strategy.{strategy_type} is required when type is '{strategy_type}'"
            )

        # Check for common mistakes in filterNode
        if strategy_type == "filterNode":
            filter_node = strategy.get("filterNode")
            if filter_node and isinstance(filter_node, dict) and "type" in filter_node:
                errors.append("filterNode should NOT have a type field (type goes in strategy)")

    # Check effect
    effect = logic.get("effect")
    if not effect or not isinstance(effect, dict):
        errors.append("effect is required and must be an object")
    else:
        if effect.get("type") != "v2":
            errors.append(f"effect.type must be 'v2', got: {effect.get('type')}")

        v2 = effect.get("v2")
        if not v2 or not isinstance(v2, dict):
            errors.append("effect.v2 is required")
        elif not v2.get("outputAndVersion"):
            errors.append("effect.v2.outputAndVersion is required")

    return ValidationResult(
        valid=len(errors) == 0,
        errors=errors,
        warnings=warnings if warnings else [],
    )


def get_strategy_type(logic: Any) -> Optional[str]:
    """Extract the strategy type from rule logic."""
    if not logic or not isinstance(logic, dict):
        return None
    strategy = logic.get("strategy")
    if not strategy or not isinstance(strategy, dict):
        return None
    return strategy.get("type")


def extract_object_type_id(logic: Any) -> Optional[str]:
    """Extract the object type ID from rule logic (works with any node type)."""
    if not logic or not isinstance(logic, dict):
        return None

    strategy = logic.get("strategy")
    if not strategy or not isinstance(strategy, dict):
        return None

    # Try each node type
    node_types = ["filterNode", "windowNode", "aggregationNode"]

    for node_type in node_types:
        node = strategy.get(node_type)
        if node and isinstance(node, dict):
            node_input = node.get("nodeInput")
            if node_input and isinstance(node_input, dict):
                source = node_input.get("source")
                if source and isinstance(source, dict):
                    object_type_id = source.get("objectTypeId")
                    if object_type_id:
                        return object_type_id

    return None


def extract_workflow_rid(logic: Any) -> Optional[str]:
    """Extract the workflow RID from rule logic."""
    if not logic or not isinstance(logic, dict):
        return None
    return logic.get("workflowRid")


def validate_workflow_rid(logic: Any, expected_rid: str) -> ValidationResult:
    """Validate that workflowRid matches expected value."""
    errors: list[str] = []
    actual_rid = extract_workflow_rid(logic)

    if not actual_rid:
        errors.append("workflowRid not found in rule logic")
    elif actual_rid != expected_rid:
        errors.append(f"workflowRid mismatch: expected '{expected_rid}', got '{actual_rid}'")

    return ValidationResult(valid=len(errors) == 0, errors=errors)


def validate_object_type(logic: Any, expected_id: str) -> ValidationResult:
    """Validate that objectTypeId matches expected value."""
    errors: list[str] = []
    actual_id = extract_object_type_id(logic)

    if not actual_id:
        errors.append("objectTypeId not found in rule logic")
    elif actual_id != expected_id:
        errors.append(f"objectTypeId mismatch: expected '{expected_id}', got '{actual_id}'")

    return ValidationResult(valid=len(errors) == 0, errors=errors)
